copyuntil(out::IO, in::IO, delim)

stevengj opened the pull request:
This PR defines and exports  ~~a new function `readuntil!(s::IO, buffer::AbstractVector{UInt8}, delim)`~~ new functions:
```jl
copyuntil(out::IO, s::IO, delim; keep=false)
copyline(out::IO, s::IO; keep=false)
```
that read data from `s` into ~~`buffer` in-place (resized if needed)~~ the `out` stream until `delim` is read/written or the end of the stream is reached.

The PR was inspired by [this post](https://discourse.julialang.org/t/is-julia-well-suited-for-string-manipulation/92943/8?u=stevengj) from @jakobnissen: the goal is to make it easier implement an allocation-free `eachline` iterator.   This can be done in a package, given `readuntil` with an `IOBuffer`, using the [`StringViews.jl` package](https://github.com/JuliaStrings/StringViews.jl) to return a string view of the in-place `buffer` on each iteration.

The reason it seemed like this needed a `Base` function, instead of living completely in a package, is that `readuntil` relies on a [low-level `jl_readuntil` C function](https://github.com/JuliaLang/julia/blob/428d242f9b8808e48905c7e82dede0e605f163cb/src/sys.c#L261-L317) that would be difficult to replicate in a package.  To obtain comparable performance, it seems like we need an analogous `jl_readuntil_buf` method (implemented in this PR) and a corresponding Julia API.

Moreover, relatively little new code was required because many of the existing `readuntil` methods used an `IOBuffer` internally, so it was merely a matter of refactoring and exporting this functionality.   Also, we already had an optimized [ios_copyuntil function](https://github.com/JuliaLang/julia/blob/8a9589d5a5d9f6bbc3dd8cbcdfb93fa03527c796/src/support/ios.c#L832-L859) for copying between IOStreams, which can now be exported in the new API.

To do:
- [x] Fix bootstrapping failures
- [x] Benchmarks.  (Maybe it is faster just to read the file in 4k blocks into a buffer with `readbytes!` and then return StringViews on top of that?   This could happen completely in a package.  It's a *lot* easier to use something like `readuntil!`, however.)
- [x] Tests
- [x] More docs
- [x] NEWS
- [x] Fixes and tests for new `out::IO` variant
- [x] add `readline(out::IO, in::IO)` too?
- [x] more tests for `readline(out::IO, in::IO)` methods
- [x] more benchmarks and optimization

Before I do much more work on this, what do people think?

stevengj committed:
in-place `readuntil!`

jakobnissen commented:
I think it looks great, especially if it's a performance boon. Thanks for working on this.
Do Julia IOStreams have their own buffer? Otherwise I could imagine having to interact with the OS for every line might be quite slow.

stevengj commented:
Yes, they do have an internal buffer, though I'm not sure how large it is.  As I mentioned, the alternative is to do some manual buffering, e.g. reading a file in 4k chunks similar to #42225.  

Of course, the advantage of manual buffering ala #42225 is that it can be implemented in pure Julia code in a package.   On the the other hand, the big *disadvantage* of manual buffering is that it is problematic for non-seekable streams (e.g. pipes) — if you want to read only a few lines, and then do something else with the remainder of the file, the user wants to be able to read from the end of the last line, not from the end of the last buffer chunk.

stevengj committed:
docstring tweaks

stevengj committed:
fix bootstrapping

stevengj committed:
bugfix

stevengj committed:
more general delimiters

stevengj committed:
fix argument order

stevengj committed:
add at least 128 bytes on resize, in case caller starts with an empty buffer

stevengj committed:
bug fixes and improvements in jl_readuntil_buf

stevengj commented:
I ran some benchmarks on `master` with the current code in this PR.  My benchmark was to open `io = open("HISTORY.md", "r")`, using the `HISTORY.md` file in the julia repo, and count the length of each line:

* **old** — `sum(length, eachline(io))` — `860.583 μs (14091 allocations: 623.17 KiB)`
* **this PR** — using `readuntil!` with a preallocated buffer and `StringViews` to wrap the buffer in a copy-free string: `671.451 μs (0 allocations: 0 bytes)`
* using `readuntil!` with a preallocated buffer and `StringViews` to wrap the buffer in a copy-free string, but *without* the low-level `jl_readline_buf` function: `18.419 ms (0 allocations: 0 bytes)`
* `sum(length, eachline(io))` with a `BufferedInputStream`: `3.781 ms (31347 allocations: 1.97 MiB)`
* `readuntil!` with a preallocated buffer and a `StringView` on a `BufferedInputStream`: 1.144 ms (0 allocations: 0 bytes)

So, only about a 30% speed improvement by eliminating the `String` allocations.  Decent but not overwhelming.  Still satisfying to see `0 allocations`, however.

If this PR is merged, it would be good to update BufferedStreams.jl to add a specialized `readuntil!` method for `BufferedInputStream` that just copies bytes in a big chunk directly from the buffer rather than the fallback of reading bytes one-by-one.   (If you don't use `memcpy`, you lose.)  Hopefully this will improve the speed even further.

-----------
Benchmark code
```jl
using StringViews, BenchmarkTools, BufferedStreams

io = open("HISTORY.md", "r")

@show sum(length, eachline(seekstart(io), keep=true))
@btime sum(length, eachline(seekstart($io), keep=true))

function doit!(io, buf, delim)
    s = 0
    while !eof(io)
        n = readuntil!(io, buf, delim)
        s += length(StringView(@view buf[1:n]))
    end
    return s
end

buf = Array{UInt8}(undef, 1024)
@show doit!(seekstart(io), buf, '\n')
@btime doit!(seekstart($io), $buf, '\n')

bio = BufferedInputStream(seekstart(io))

@show sum(length, eachline((seekstart(bio); bio), keep=true))
@btime sum(length, eachline((seekstart($bio); $bio), keep=true))

@show doit!((seekstart(bio); bio), buf, '\n')
@btime doit!((seekstart($bio); $bio), $buf, '\n')

close(bio)
```

stevengj committed:
add readuntil! to manual

stevengj committed:
optimized IOBuffer readuntil

stevengj commented:
I added an optimized `readuntil!` for `IOBuffer`, and modified `readuntil(::IOBuffer, ...)` to use it.  It seems to be a net win even for traditional `eachline` usage that allocates a string on each call.  e.g.
```jl
iob = IOBuffer(read("HISTORY.md"))
@btime sum(length, eachline(seekstart($iob), keep=true))
```
previously gave `1.622 ms (23488 allocations: 1.96 MiB)` and now gives `1.162 ms (23484 allocations: 1.96 MiB)`.

If I use an `IOBuffer` with `readuntil!` and a pre-allocated buffer with `StringView` as above, it gives `697.035 μs (0 allocations: 0 bytes)`.

stevengj committed:
tests, fixes

stevengj committed:
NEWS

stevengj commented:
Should be ready for review.

stevengj committed:
bugfix

stevengj commented:
Other related threads:

* https://discourse.julialang.org/t/bad-performance-of-eachline-on-stdin/55701/34
* https://discourse.julialang.org/t/slow-gzip-streaming-in-julia-but-not-in-python/56346
* https://discourse.julialang.org/t/package-to-read-process-lines-without-new-allocations/92448

stevengj commented:
cc @rickbeeloo, author of the https://github.com/rickbeeloo/ViewReader package.

stevengj commented:
I was thinking about it some more, and arguably if you want to have maximum performance for an in-place `eachline`-like iteration, then you maybe want to read from the stream in large blocks (containing many linesa) ala BufferedStreams and then *don't copy* at all if possible — just return a `StringView` directly into the buffer.  (Only a little data motion is required when you get to the end of the buffer, and need to move the final line fragment back to the beginning to read more.)   I think this is what ViewReader.jl does?

From that perspective, however, a `readuntil!` API is unnecessary.

jakobnissen commented:
True, and indeed something like Rust's std::Io::BufReader might be nice to have. However, how would you keep track of whether it's safe to shift data in the buffer, given that StringViews may hold a reference to any data in the buffer?

rickbeeloo commented:
I also agree reading bigger chunks would do better, especially for those reading from HDDs

> Only a little data motion is required when you get to the end of the buffer, and need to move the final line fragment back to the beginning to read more

This works as long as the "to be finished line" is shorter than the currently allocated buffer. If not it might again not find the newline. That would require some extra logic to see if the newline is found yet and otherwise still increase the buffer and extend. Which, in the worst case would give a buffer ~2x the longest line in the file. I didn't really spend time optimizing that, instead, I allocated two buffers (each being the max line length - given by the user) then flip them around like you said and warn if no newline is found after

> Though it may still be more convenient for some applications, especially if you want to read other data in between reading lines

`readuntil!` will still also still be increasingly faster for longer lines (not like in `HISTORY.md`)

stevengj commented:
Marking for `triage` since it would be good to get some feedback from core devs on whether this API is desired.

stevengj commented:
I was thinking about this some more in connection with #48625, and I'm starting to feel that a better API would be:
```jl
readuntil(out::IO, in::IO, delim)
```
i.e. add an optional `out::IO` argument instead of a buffer.   Advantages:

* Includes functionality of a buffer, since you can pass a preallocated `IOBuffer`.
* Greater functionality because now you can output to arbitrary streams.
* Little new code since much of the code already uses `IOBuffer` internally; just needs to be refactored.
* We already have an optimized [`ios_copyuntil` function](https://github.com/JuliaLang/julia/blob/8a9589d5a5d9f6bbc3dd8cbcdfb93fa03527c796/src/support/ios.c#L832-L859) for copying between `IOStream`s. The new API would let us export this.

stevengj committed:
readuntil(out::IO, ...) instead of readuntil!

stevengj commented:
Updated to use the new `readuntil(out::IO, in::IO, delim)` API.   Not tested yet and probably slightly broken, but should give a good sense of what the code will look like.

fredrikekre commented:
There is already `write(out::IO, in::IO)`, which is more or less `readuntil(out, in , eof)`. Perhaps this should also be a `write` method?

stevengj commented:
I think it's also more discoverable to stick to the convention that you can add an `out::IO` method as the first argument of a "string" function to make it write to a buffer instead of returning a string.  That way you don't have to go hunting around for "what is the buffer equivalent of `foo(...)`", since it's always `foo(io, ...)`.

For example, we already have `join(...)` and `join(io, ...)`, and I think we should also add an `io` argument to `replace` (#48625), and probably to other functions as well.  They can't all be methods of `write` or `print`.

stevengj committed:
bugfixes

stevengj committed:
rm stray semicolon

stevengj committed:
readline(out, ...)

JeffBezanson commented:
This is great functionality, triage is :+1: . However, we do like the idea of calling it `copy` and/or `copyuntil` since it both reads and writes. This can be merged when a name is agreed on, but it seems to me with the code here we should also be able to get versions that (1) copy the entire stream and (2) copy a set number of bytes as well.

stevengj commented:
`copyuntil` seems fine to me.

stevengj committed:
readuntil -> copyuntil

stevengj commented:
Renamed to `copyuntil`.

Functionality for copying entire streams, e.g. `copy(outstream, instream)`, or a fixed number of bytes, seem best left to a separate PR.

stevengj commented:
For some reason it's now slower than `readline`, even though it is allocation free:
```jl
using StringViews, BenchmarkTools
io = open("HISTORY.md", "r")

@show sum(length, eachline(seekstart(io), keep=true))
@btime sum(length, eachline(seekstart($io), keep=true))

buf = IOBuffer(sizehint=1024)
function doit!(io, buf, delim)
    s = 0
    while !eof(io)
        seekstart(buf).size = 0
        n = copyuntil(buf, io, delim, keep=true)
        s += length(StringView(buf))
    end
    return s
end
@show doit!(seekstart(io), buf, UInt8('\n'))
@btime doit!(seekstart($io), $buf, UInt8('\n'))
```
gives
```
sum(length, eachline(seekstart(io), keep = true)) = 371831
  880.198 μs (14092 allocations: 622.91 KiB)
doit!(seekstart(io), buf, UInt8('\n')) = 371831
  941.902 μs (0 allocations: 0 bytes)
```
Not sure what gives.  Any ideas?

stevengj committed:
add manual entries for copyuntil and copyline

stevengj commented:
Weird, I'm getting an test error on Windows:
```
nested task error: IOError: write: end of file (EOF)
--
  | Stacktrace:
  | [1] uv_write(s::Base.PipeEndpoint, p::Ptr{UInt8}, n::UInt64)
  | @ Base .\stream.jl:1066
  | [2] unsafe_write(s::Base.PipeEndpoint, p::Ptr{UInt8}, n::UInt64)
  | @ Base .\stream.jl:1120
  | [3] write
  | @ .\strings\io.jl:244 [inlined]
  | [4] macro expansion
  | @ C:\buildkite-agent\builds\win2k22-amdci6-6\julialang\julia-master\julia-9f5194be53\share\julia\test\read.jl:55 [inlined]
```
from [the TCPsocket test `io` pipe](https://github.com/JuliaLang/julia/blob/9f5194be539fe6a5fbd3502827c9078a18612c3d/test/read.jl#L55).

Am I supposed to call `cleanup()` more often between the `read` tests?  I'm not sure why that function is called so often…

stevengj committed:
try calling cleanup() more often in test for Windows

stevengj commented:
CI failure looks unrelated (`nested task error: rmprocs: pids [2] not terminated after 30 seconds.` in Distributed `cluster.jl` tests.

stevengj committed:
use _unsafe_take and a few other tweaks

stevengj committed:
bugfix: missing ensureroom

stevengj commented:
So nice to see green CI again!

stevengj commented:
bump … anything needs to be done here before merging?

Robert-j7 commented:
Will this be merged into 1.9?

Seelengrab commented:
Feature freeze for 1.9 has been quite a while ago.

stevengj commented:
Rebased.

stevengj commented:
![light is green](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYjU3MmMwOGI2NGI0NzQ0ZWU3YzYzMmY4MWNmZGM5NmExMjM3MzNmOSZjdD1n/3KlUmxJN7WUetKHknh/giphy.gif)

Robert-j7 commented:
If nothing else to do, please merge this.

rfourquet commented on base/io.jl:
```diff
@@ -499,11 +499,15 @@ read!(filename::AbstractString, a) = open(io->read!(io, a), convert(String, file
 Keyword argument `keep` controls whether the delimiter is included in the result.
 The text is assumed to be encoded in UTF-8.
 
+Returns a `String` if `delim` is an `AbstractChar` or a string
+or otherwise returns a `Vector{typeof(delim)}`.   See also [`copyuntil`](@ref)
+to instead write in-place to another stream (which can be a preallocated [`IOBuffer`](@ref)).
```
Unless the `s` was intentional:
```suggestion
Return a `String` if `delim` is an `AbstractChar` or a string
or otherwise return a `Vector{typeof(delim)}`.   See also [`copyuntil`](@ref)
to instead write in-place to another stream (which can be a preallocated [`IOBuffer`](@ref)).
```

stevengj replied:
The `s` was intentional — grammatically, I was thinking of the implicit subject being `copyuntil` (or "it").   i.e. "`copyuntil` returns a …".     It's descriptive, not imperative.

stevengj replied:
Interestingly, the descriptive tense "Returns …" is used 45 times in the Julia docs, while the imperative tense "Return …" is used 288 times.

stevengj replied:
See #50301 if you want to have a longer discussion.

rfourquet commented on base/io.jl:
```diff
@@ -530,6 +566,11 @@ false (as it is by default), these trailing newline characters are removed from
 line before it is returned. When `keep` is true, they are returned as part of the
 line.
 
+Returns a `String`.   See also [`copyline`](@ref) to instead write in-place
```
```suggestion
Return a `String`.   See also [`copyline`](@ref) to instead write in-place
```

stevengj committed:
Update base/io.jl

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>

stevengj committed:
Update base/io.jl

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>

nalimilan commented:
Once this is merged, would it make sense to also have `copyavailable` to offer an in-place version of `readavailable`? That could be useful to define a more efficient fallback for `write`, which currently makes a super wasteful copy:
https://github.com/JuliaLang/julia/blob/de73c26fbff61d07a38c9653525b530a56630831/base/io.jl#L752-L755
That would help with performance issues such as https://github.com/JuliaData/CSV.jl/issues/1022.

rickbeeloo commented:
@stevengj Thanks that you are still working on this one, would be awesome to have it!

stevengj commented:
Unfortunately, looks like this didn't get merged in time for the 1.10 feature freeze.  Oh well, it can wait until 1.11 — rebased.

mkitti commented:
@giordano Does getting this green involve similar magic as https://github.com/JuliaLang/julia/pull/48625#issuecomment-1602564337?

giordano commented:
There's a segfault in AddressSanitizer

stevengj commented:
I'm having trouble seeing how the Asan segfault (in `src/flisp/print.c`) could be due to this PR.

vtjnash commented on base/iostream.jl:
```diff
@@ -443,11 +443,46 @@ end
+        end
+        (eof(s) || len == out.maxsize) && break
+        len = min(2len + 64, out.maxsize)
+        resize!(d, len)
```
This seems like it might not quite respect maxsize. Should it use `ensureroom` instead? The `ensureroom` call at the top also seems wrong, since if someone new the `out` stream was already sized correctly for a smaller read, that will forceably try to reallocate it.

stevengj replied:
I'm not sure what you mean by "not copying the last chunk".  By the time it hits the `len == out.maxsize` branch, `jl_readuntil_buf` has already copied `out.maxsize` bytes into the buffer.   The `jl_readuntil_buf` never reads a chunk exceeding `maxsize`, because the number of bytes that it reads is bounded by `len - ptr + 1`.

`ensureroom` seems like it has quite a few extra checks that aren't needed in the inner loop here given the `ensureroom` call at the top.

I guess it wouldn't hurt to change the `ensureroom` call at the top to something like `ensureroom(out, iszero(io.size) ? 16 : 0)`

vtjnash replied:
Ah, I guess that makes sense. The checks were a bit distant so it wasn't immediately obvious that was being ensured.

Yes, I think checking `isempty` first makes sense. I realized now that `ensureroom` already truncates the request to maxsize, so there is no issue with calling that.

stevengj replied:
should be resolved by #50485

JeffBezanson commented on base/io.jl:
```diff
@@ -816,36 +899,38 @@ end
 end
+readuntil(s::IO, delim::T; keep::Bool=false) where T =
+    _copyuntil(Vector{T}(), s, delim, keep)
+readuntil(s::IO, delim::UInt8; keep::Bool=false) =
```
This overwrites the definition on line 525, giving a warning during sysimage build.

giordano replied:
Is there a way to catch those warnings and turn them into errors on CI when building the sysimage? I guess we don't want this kind of situations.

stevengj replied:
Will file a PR to fix this shortly.
